### PR TITLE
Add atom mapping filter disallowing flipped planar torsions in core

### PR DESCRIPTION
**Pull request by mcwitt**
Adds an atom mapping filter removing mappings that flip a planar torsion (having a double or aromatic central bond) is flipped.

### CODE REVIEW

File: timemachine/fe/chiral_utils.py [hunk 1]
```diff
@@ -342,3 +347,44 @@ def find_chiral_bonds(mol):
...
+        conf = get_romol_conf(mol)
+        graph = convert_to_nx(mol)
+        idxs = {canonicalize_bond(tuple(idxs)) for idxs in enumerate_simple_paths(graph, 4)}
+
+        planar_torsions = dict()
+        for i, j, k, l in idxs:
+            bond_type = mol.GetBondBetweenAtoms(j, k).GetBondType()
+            if bond_type == BondType.DOUBLE or bond_type == BondType.AROMATIC:
+                volume = torsion_volume(conf[i], conf[j], conf[k], conf[l])
+                planar_torsions[(i, j, k, l)] = np.sign(volume)
```
Comment (proteneer) [inline timemachine/fe/chiral_utils.py:379]: Possible to also special case amide bonds?

``` python
from rdkit import Chem

def find_canonical_amide_bonds(mol):
    query = Chem.MolFromSmarts("[NX3]-[CX3](=[OX1])")
    res = mol.GetSubstructMatches(query)
    
    amide_bonds = []
    for i,j,k in res:
        bonds.append(sorted([i,j]))
    
    return amide_bonds
    
mol = Chem.MolFromSmiles("CN1CCC(CC1=O)NC(C)=O")
find_canonical_amide_bonds(mol)
```

Comment (mcwitt) [reply timemachine/fe/chiral_utils.py]: Updated to include torsions with amide central bonds in the check in https://github.com/proteneer/timemachine/pull/1229/commits/53ae67da5c6239ddab18a7ff6012843bccea890d

File: timemachine/md/minimizer.py [hunk 2]
```diff
@@ -551,3 +551,17 @@ def val_and_grad_fn_bfgs(x_local_flattened):
...
+
+
+def replace_conformer_with_minimized(mol: Chem.rdchem.Mol, ff: Forcefield):
+    top = topology.BaseTopology(mol, ff)
+    system = top.setup_end_state()
+    val_and_grad_fn = get_val_and_grad_fn(system.get_U_fns(), 1e9 * np.eye(3))
+    xs = get_romol_conf(mol)
+    all_idxs = np.arange(mol.GetNumAtoms())
+    xs_opt = local_minimize(xs, val_and_grad_fn, all_idxs, verbose=False)
+    conf = Chem.Conformer(mol.GetNumAtoms())
```
Comment (badisa) [inline timemachine/md/minimizer.py:563]: Would be good to use `set_romol_conf` here

File: timemachine/fe/chiral_utils.py [hunk 3]
```diff
@@ -342,3 +347,56 @@ def find_chiral_bonds(mol):
...
+def setup_find_flipped_planar_torsions(mol_a, mol_b):
+    def enumerate_planar_torsions(mol):
+        conf = get_romol_conf(mol)
+        graph = convert_to_nx(mol)
+        idxs = {canonicalize_bond(tuple(idxs)) for idxs in enumerate_simple_paths(graph, 4)}
+        amide_bonds = find_canonical_amide_bonds(mol)
+
+        planar_torsions = dict()
+        for i, j, k, l in idxs:
+            if (j, k) not in amide_bonds:
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/fe/chiral_utils.py]: Good catch. Fixed in https://github.com/proteneer/timemachine/pull/1229/commits/f1e5839896c2db658101a3cd3d16404e2461f31d.

File: timemachine/md/minimizer.py (hunk 2 above)
Comment (mcwitt) [reply timemachine/md/minimizer.py]: Thanks, that's much cleaner. Fixed in https://github.com/proteneer/timemachine/pull/1229/commits/2037d231e40767a8f34f21def108c2d4b54ce91e.

File: timemachine/fe/chiral_utils.py [hunk 4]
```diff
@@ -342,3 +347,56 @@ def find_chiral_bonds(mol):
             chiral_bonds.add(tuple(sorted([match[0], match[1]])))
 
     return chiral_bonds
+
+
+def find_canonical_amide_bonds(mol):
```
Comment (mcwitt) [inline timemachine/fe/chiral_utils.py:352]: This seems to have a lot of overlap with `find_chiral_bonds` defined above, but not sure if it makes sense to combine them.

Comment (proteneer) [reply timemachine/fe/chiral_utils.py:352]: Let's leave them separate for now. The `find_chiral_bonds` code is probably unused/deprecated.

File: timemachine/fe/atom_mapping.py [hunk 5]
```diff
@@ -343,19 +349,31 @@ def _get_cores_impl(
...
-    else:
-        filter_fxn = lambda core: True
+        filter_fxns.append(chiral_filter)
+
+    if disallow_planar_torsion_flips:
+        find_flipped_planar_torsions = setup_find_flipped_planar_torsions(mol_a, mol_b)
+
+        def planar_torsion_flip_filter(trial_core):
+            flipped = find_flipped_planar_torsions(trial_core)
+            passed = flipped == []
```
Comment (badisa) [inline timemachine/fe/atom_mapping.py:370]: nit:

```suggestion
            passed = len(flipped) == 0
```

**Review by badisa — APPROVED** [review]
Good solution to this issue.

It might be nice to add some of the plots you had posted offline to the description of the PR, mostly for being able to refer to the impact of this PR in the future.

File: timemachine/md/minimizer.py [hunk 6]
```diff
@@ -551,3 +552,27 @@ def val_and_grad_fn_bfgs(x_local_flattened):
         warnings.warn(f"WARNING: Energy did not decrease: u_0: {u_0:.3f}, u_f: {U_final:.3f}", MinimizationWarning)
 
     return x_final
+
+
+def replace_conformer_with_minimized(mol: Chem.rdchem.Mol, ff: Forcefield):
+    """
+    Replace the first conformation of the given mol with a conformation minimized with respect to the given forcefield.
+
+    Intended to be a drop-in replacement for :py:func:`rdkit.Chem.AllChem.EmbedMolecule` that allows specifying the
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/md/minimizer.py]: Ah, good point. Fixed in https://github.com/proteneer/timemachine/pull/1229/commits/c853eedac3ea18196b33e96bd8a682c98c7638a8

File: timemachine/fe/atom_mapping.py (hunk 5 above)
Comment (mcwitt) [reply timemachine/fe/atom_mapping.py]: Changed in https://github.com/proteneer/timemachine/pull/1229/commits/fe75bf3b84217bdace0d630320aa824b90cd711c

**Review by proteneer — APPROVED** [review]

**Pull request merged**